presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java

```diff
@@ -849,4 +969,61 @@ public FunctionMap(FunctionMap map, Iterable<? extends SqlFunction> functions)
             return functions.get(name);
         }
     }
+
+    private class ApplicableFunction
+    {
+        private final Signature declaredSignature;
+        private final Signature boundSignature;
+
+        private ApplicableFunction(Signature declaredSignature, Signature boundSignature)
+        {
+            this.declaredSignature = declaredSignature;
+            this.boundSignature = boundSignature;
+        }
+
+        public Signature getDeclaredSignature()
+        {
+            return declaredSignature;
+        }
+
+        public Signature getBoundSignature()
+        {
+            return boundSignature;
+        }
+
+        /**
+         * One method is more specific than another if invocation handled by the first method could be passed on to the other one
+         */
+        public boolean isMoreSpecificThan(ApplicableFunction other)
+        {
+            List<Type> resolvedTypes = resolveTypes(boundSignature.getArgumentTypes(), typeManager);
+            Optional<BoundVariables> boundVariables = new SignatureBinder(typeManager, other.getDeclaredSignature(), true)
+                    .bindVariables(resolvedTypes);
+            return boundVariables.isPresent();
+        }
+
+        public boolean isInSameFamilyAs(ApplicableFunction other)
```

Comment:
The simplest possible definition is defined by this method:
The family is the set (class) of ApplicableFunctions. The functions are in the same family when any of them is more specific then the other.
I've added this explanation as comment.

Generally what ApplicableFunctions defines is simple rule of ordering where some elements are comparable and some are not. The family is set of elements that are comparable. The set is in fact a class, as the defined relation is transitive.